### PR TITLE
completed week 4 assessment

### PR DESCRIPTION
**Pull request by Lilmcg**
Hi Austin. I'm stumped why two of my Ruby coding answers would not work. I guess I'll find out this morning! Hope you had a great time at the wedding!

### CODE REVIEW

File: code_challenges.rb [hunk 1]
```diff
@@ -21,6 +32,18 @@
...
+  no_vowels.delete('aeiouAEIOU')
+end
+
+p no_vowels(album1)
+p no_vowels(album2)
+p no_vowels(album3)
+
+ # // the above method does not work. I receive an error message.
+ # // the simple code below works, however it's not set up as a true method that can be run repeatedly with a new argument each time...
+
```
Comment (WalkerTexas-coder) [inline code_challenges.rb:45]: Ah got it, so something to keep in mind (and I have seen this happen in your JS challenges in the past too (not trying to rip on you, just giving a heads up)). 

Generally speaking we don't (can't?) call on the method or function name inside the function or method itself. 

so here on line 36 you are going inside of the method called **_no_vowels_** and there, inside the method, you are calling on **_no_vowels_**. 

**_no_vowels_** is being defined by you the developer so it doesn't know what to do with **_no_vowels_** inside its own definition. 

Now there are times to call on a function inside of itself: thats called recursion. We don't do a whole lot of recursion in class as its a concept more present in Functional Programming.

File: code_challenges.rb [hunk 2]
```diff
@@ -31,3 +54,18 @@
...
+  else
+    "#{string} is not a palindrome"
+  end
+end
+
+p palindrome(is_palindrome1)
+p palindrome(is_palindrome2)
+p palindrome(is_palindrome3)
+
+# // the above method does not work. All 3 outputs state "...is not a palindrome". It doesn't seem to be getting past the capitalization in Racecar and Rotator since those should return with the output "...is a string"
```
Comment (WalkerTexas-coder) [inline code_challenges.rb:71]: great observation! And you are totally correct its getting hung up on the capitalization! If you follow that down the rabbit hole you can make a small change to your evaluation logic and ...
```ruby
def palindrome(string)
  if string.downcase == string.downcase.reverse
    "#{string} is a palindrome"
  else
    "#{string} is not a palindrome"
  end
end
``` 

Boom you got it!

File: interview-questions.md [hunk 3]
```diff
@@ -7,52 +7,52 @@ Then, research the question to expand on your answer. Even if you feel you have
 
 1. What are props in React?
 
-  Your answer:
+  Your answer: Props is a keyword in React that let's you pass info across components, but only in one direction. The info must pass down from the parent to the child.
 
-  Researched answer:
+  Researched answer: Props stands for properties. It is a keyword in React that let's you pass info across components, but only in one direction. The info must pass down from the parent to the child. Props come mostly from state and they cannot be updated. They are a snapshot of state. It's a great way to make the data much more available and they make the app dynamic so that any updates to data will not require major re-coding in all areas.
```
Comment (WalkerTexas-coder) [inline interview-questions.md:12]: Amazing answer!

File: interview-questions.md [hunk 4]
```diff
@@ -7,52 +7,52 @@ Then, research the question to expand on your answer. Even if you feel you have
...
 
 
 
 3. What is object-oriented programming? How is it different than functional programming?
 
-  Your answer:
+  Your answer: OOP is a programming model that relies on classes and objects versus function and logic.
 
-  Researched answer:
+  Researched answer: OOP is a programming model that relies on classes and objects versus function and logic. It creates object based elements that are held in hierarchy. Methods belong to, and manipulate the objects. OOP is good when you have a fixed set of operations that won't need to be changed, but when they do, you can add new classes which work off of existing methods (the original classes are left alone). The original classes are generally left alone but the data can be changed. OOP is an easier way to add new data and functions. Functional programming focuses on using functions that perform a specific task. Importance is given to the functions more so than the data. New functions are added that use the existing data types (data in the functions cannot be changed after they have been created), and the original functions are left alone.
```
Comment (WalkerTexas-coder) [inline interview-questions.md:28]: Well done! Yeah these rabbit holes go deep!

File: interview-questions.md [hunk 5]
```diff
@@ -7,52 +7,52 @@ Then, research the question to expand on your answer. Even if you feel you have
...
 
 
 
 4. What is the difference between a Float and an Integer in Ruby?
 
-  Your answer:
+  Your answer: An integer in Ruby is a whole number. A float in Ruby is a partial number.
 
-  Researched answer:
+  Researched answer: Both integer and float are numerical data in Ruby. An integer in Ruby is a whole number and it does not have a decimal point. A float in Ruby is a partial number that contains a decimal point. Float is used when you need a more precise number.
```
Comment (WalkerTexas-coder) [inline interview-questions.md:36]: Well done!

File: code-challenges.test.js [hunk 6]
```diff
@@ -49,6 +89,28 @@ var testArray1 = [3, 7, 10, 5, 4, 3, 3]
...
+
+// Create a function arrayNew that takes in 2 arrays as parameters
+// Merge arrays together using .concat
+// Use .filter to iterate through the combined array and find whether that value is present in the combined array
+  // Use .indexOf to filter any values that occur multiple times and return non-duplicates
+
+const arrayNew = (array1, array2) => {
+  let combineArray = array1.concat(array2)
+  return combineArray.filter((value, index, array) => {
+    return array.indexOf(value) === index
```
Comment (WalkerTexas-coder) [inline code-challenges.test.js:110]: AWESOME!

File: code-challenges.test.js [hunk 7]
```diff
@@ -20,26 +20,66 @@ var colors1 = ["purple", "blue", "green", "yellow", "pink"]
...
+// Ran test. Test failed. Error: getMinMax is not defined
 
 // b) Create the function that makes the test pass.
 
+// Create a function getMinMax that take in an array
+// Use the Math object and its .min and .max functions to return the min and max of the array. Use the spread operator to convert the array into a list of individual values and then min and max can work on those values
+
+const getMinMax = (array) => {
+  return [Math.min(...array), Math.max(...array)]
+}
```
Comment (WalkerTexas-coder) [inline code-challenges.test.js:78]: This is so clean! Love one line solutions

File: code-challenges.test.js [hunk 8]
```diff
@@ -20,26 +20,66 @@ var colors1 = ["purple", "blue", "green", "yellow", "pink"]
...
 // Expected output example (can be a different order): ["saffron", "aquamarine", "periwinkle", "indigo", "ochre"]
 
+describe ("shuffleArray", () => {
+  test("given an array, remove the first item and shuffle the remaining", () => {
+    expect(shuffleArray(colors1)).toContain("yellow", "blue", "pink", "green")
+    expect(shuffleArray(colors2)).toContain("saffron", "aquamarine", "periwinkle", "indigo", "ochre")
+  })
+})
 
+// Ran test. Test failed. Error: shuffleArray is not defined.
```
Comment (WalkerTexas-coder) [inline code-challenges.test.js:30]: Perfect

File: code-challenges.test.js [hunk 9]
```diff
@@ -20,26 +20,66 @@ var colors1 = ["purple", "blue", "green", "yellow", "pink"]
...
+       var randomArray = Math.floor(Math.random() * (i + 1));
+       [array[i], array[randomArray]] = [array[randomArray], array[i]]
+   }
+    return array
+}
+
+// console.log(shuffleArray(colors2))
+// console.log(shuffleArray(colors1))
+
+// Ran test. Test passed!
```
Comment (WalkerTexas-coder) [inline code-challenges.test.js:51]: Love it!

**Review by WalkerTexas-coder — APPROVED** [review]
Lillian, great work! The logic you employed in the js challenges is solid. You've really nailed down some of the finer points of those tools and its great to see. I replied specifically to the ruby issues you ran into below, you are 95% the way there on both. And you did a great job deep divin ginto the interview questions. I want to encourage you to share a bit in an interview setting about a time you interacted with "the subject" thats being asked. It conveys familiarity and understanding and is a generally good interview tactic! Great work Lillian-- This PR is approved!

**Comment by Lilmcg** [comment]
Thanks Austin! Can’t believe the silly mistake trying to run the .delete on the function instead of my string in Ruby🤦🏼‍♀️

Lil

**Pull request merged**